### PR TITLE
Examples vanilla pnp

### PR DESCRIPTION
**Pull request by matthieutrs**
PRing a branch by @samuro95 with many changes, including:
- refactoring of the class builders;
- adding PnP examples;

Docs not yet passing, currently fixing.

### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.md](../../CHANGELOG.md) (as applicable) (to come).

### CODE REVIEW

File: deepinv/training_utils.py [hunk 1]
```diff
@@ -255,56 +251,43 @@ def test(
...
-
-                    if save_images:
-                        for img, name_im in zip(imgs, name_imgs):
-                            im_save(
-                                save_folder_G / (name_im + "_" + str(i) + ".png"),
-                                torch2cpu(img),
-                            )
-                    if plot_images:
-                        plot(imgs, titles=name_imgs)
+                    if len(y.shape) == 4:
```
Comment (tachella) [inline deepinv/training_utils.py:284]: nice fix

File: deepinv/unfolded/deep_equilibrium.py [hunk 2]
```diff
@@ -128,25 +100,42 @@ def DEQ(
...
-
-    :param str algo_name: name of the algorithm to be used. Should be either `"PGD"`, `"ADMM"`, `"HQS"`, `"PD"` or `"DRS"`.
-    :param dict params_algo: dictionary containing the parameters of the algorithm.
-    :param list trainable_params: list of trainable parameters. Default: `[]`.
-    :param deepinv.optim.data_fidelity data_fidelity: data fidelity term in the optimization problem.
-    :param F_fn: Custom user input cost function. Default: None.
-    :param g_first: whether to perform the step on :math:`g` before that on :math:`f` before or not. Default: False.
+    Function building the appropriate Unfolded architecture.
+
+    :param algo: either name of the algorithm to be used, or an iterator. If an algorithm name (string), should be either `"PGD"`, `"ADMM"`, `"HQS"`, `"CP"` or `"DRS"`.
```
Comment (tachella) [inline deepinv/unfolded/deep_equilibrium.py:105]: Should we call it iterator as in the optim builder?

File: deepinv/unfolded/unfolded.py [hunk 3]
```diff
@@ -61,36 +60,41 @@ def __init__(
...
+    iteration, data_fidelity=L2(), F_fn=None, g_first=False, beta=1.0, **kwargs
 ):
     r"""
     Function building the appropriate Unfolded architecture.
 
     :param algo: either name of the algorithm to be used, or an iterator. If an algorithm name (string), should be either `"PGD"`, `"ADMM"`, `"HQS"`, `"CP"` or `"DRS"`.
-    :param deepinv.optim.DataFidelity data_fidelity: data fidelity term in the optimization problem.
-    :param F_fn: Custom user input cost function. Default: None.
-    :param g_first: whether to perform the step on :math:`g` before that on :math:`f` before or not. Default: False.
+    :param deepinv.optim.data_fidelity data_fidelity: data fidelity term in the optimisation problem.
```
Comment (tachella) [inline deepinv/unfolded/unfolded.py:70]: This change is not correct
- The class if DataFidelity
- We use american english

File: deepinv/unfolded/unfolded.py [hunk 4]
```diff
@@ -61,36 +60,41 @@ def __init__(
...
+    :param F_fn: Custom user input cost function. default: None.
```
Comment (tachella) [inline deepinv/unfolded/unfolded.py:71]: should this be a callable type?

File: deepinv/utils/demo.py [hunk 5]
```diff
@@ -105,3 +108,31 @@ def load_degradation(name, data_dir, kernel_index=0, download=True):
...
+
+
+def load_image(
+    path=None,
+    url=None,
+    img_size=None,
+    grayscale=False,
+    resize_mode="crop",
+    device="cpu",
+):
```
Comment (tachella) [inline deepinv/utils/demo.py:120]: we can add some docstring explaining the function.
I would also rename it to load_url_image

File: deepinv/utils/plotting.py [hunk 6]
```diff
@@ -84,36 +97,77 @@ def plot(img_list, titles=None, save_dir=None, tight=True, max_imgs=4):
...
+            for r, img in enumerate(row_imgs):
+                plt.imsave(
+                    save_dir / (titles[i] + "_" + str(r) + ".png"), img, cmap="gray"
+                )
+    if show:
+        plt.show()
+
 
-    plt.show()
+def plot_curves(metrics, save_dir=None, show=True):
```
Comment (tachella) [inline deepinv/utils/plotting.py:132]: we need to document the function

File: deepinv/utils/plotting.py [hunk 7]
```diff
@@ -126,3 +180,63 @@ def wandb_imgs(imgs, captions, n_plot):
             )
         )
     return wandb_imgs
+
+
+def wandb_plot_curves(metrics, batch_idx=0, step=0):
```
Comment (tachella) [inline deepinv/utils/plotting.py:185]: add docstrings

File: deepinv/utils/plotting.py [hunk 8]
```diff
@@ -126,3 +180,63 @@ def wandb_imgs(imgs, captions, n_plot):
...
+                        keys=[f"image {j}" for j in range(batch_size)],
+                        title=f"{metric_name} batch {batch_idx}",
+                        xname="iteration",
+                    )
+                },
+                step=step,
+            )
+
+
+def plot_parameters(model, init_params=None, save_dir=None, show=True):
```
Comment (tachella) [inline deepinv/utils/plotting.py:203]: add docstrings

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
+"""
+PnP with custom optimization algorithm (Condat-Vu Primal-Dual)
+====================================================================================================
+
+This example shows how to set its own custom optimization algorithm.
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:5]: "how to define your own optimization algorithm"

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
+"""
+PnP with custom optimization algorithm (Condat-Vu Primal-Dual)
+====================================================================================================
+
+This example shows how to set its own custom optimization algorithm. 
+For example, here, we implement the Condat-Vu Primal-Dual algorithm for Single Pixel Camera (SPC) reconstruction.
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:6]: Maybe here we should explain a little bit more, that the only thing that the user has to do is define an iterator.

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+# --------------------------------------------------------------------------------
+# We use the Proximal Gradient Descent optimization algoritm.
+# The algorithm alternates between a denoising step and a gradient descent step.
+# The denoising step is performed by a DNCNN pretrained denoiser :class:`deepinv.models.dncnn`.
+
+# Logging parameters
+verbose = True
+plot_metrics = True  # compute performance and convergence metrics along the algorithm, curved saved in RESULTS_DIR
+
+# Set up the PnP algorithm parameters : the `stepsize`, `g_param` the noise level of the denoiser and `lambda` the regularization parameter. The following parameters are chosen arbitrarily.
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:175]: The following parameters can be chosen by cross-validation?

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+    in_channels=n_channels,
+    out_channels=n_channels,
+    pretrained="download",
+    train=False,
+    device=device,
+)
+prior = PnP(denoiser=denoiser)
+
+# instantiate the algorithm class to solve the IP problem.
+algo = CVIteration(data_fidelity=data_fidelity, F_fn=None, has_cost=False)
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:194]: iterator instead of algo?

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+        u = self.f_step(physics.A(2 * x - x_prev), u_prev, y, cur_params)
+
+        F = self.F_fn(x, cur_params, y, physics) if self.has_cost else None
+
+        return {"est": (x, u), "cost": F}
+
+
+class fStepCV(fStep):
+    r"""
+    PD fStep module.
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:70]: maybe we can add in latex the fstep here

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+        """
+        v = u + cur_params["stepsize"] * Ax_cur
+        return v - cur_params["stepsize"] * self.data_fidelity.prox_d(
+            v, y, 1 / (cur_params["stepsize"] * cur_params["lambda"])
+        )
+
+
+class gStepCV(gStep):
+    r"""
+    PD gStep module.
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:93]: idem f step

File: examples/plug-and-play/demo_PnP_custom_optim.py [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+)
+
+# %%
+# Evaluate the model on the problem and plot the results.
+# --------------------------------------------------------------------
+
+y = physics(x)
+x_lin = physics.A_adjoint(y)
+
+# run the model on the problem. When `return_metrics` is set to True, the model requires the ground-truth clean image ``x_gt`` and returns the output and the metrics computed along the iterations.
```
Comment (tachella) [inline examples/plug-and-play/demo_PnP_custom_optim.py:213]: is it possible to have return_metric = True and no gt?

**Review by tachella — CHANGES_REQUESTED** [review]
Great PR, lots of fixes!
I just highlighted some little extra points to fix before pushing

File: deepinv/unfolded/deep_equilibrium.py (hunk 2 above)
Comment (samuro95) [reply deepinv/unfolded/deep_equilibrium.py:105]: The input type with 'or' ?

File: examples/plug-and-play/demo_PnP_custom_optim.py (hunk 15 above)
Comment (samuro95) [reply examples/plug-and-play/demo_PnP_custom_optim.py:213]: actually yes

File: examples/unfolded/demo_LISTA.py [hunk 16]
```diff
@@ -239,73 +244,15 @@
...
-    stepsize,
-    label="init. stepsize",
-    color="b",
-    linestyle="dashed",
-)
-ax.plot(
-    np.arange(len(list_stepsize)), list_stepsize, label="learned stepsize", color="b"
+# Plotting the learned parameters.
+# ------------------------------------
+dinv.utils.plotting.plot_parameters(
```
[the remark this answers is not in the document]
Comment (samuro95) [reply examples/unfolded/demo_LISTA.py:256]: the plot_parameters fct ?

File: examples/plug-and-play/demo_PnP_custom_optim.py (hunk 15 above)
Comment (tachella) [reply examples/plug-and-play/demo_PnP_custom_optim.py]: ok, maybe we can clarify in the comment

File: examples/unfolded/demo_LISTA.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (tachella) [reply examples/unfolded/demo_LISTA.py:256]: I mean just adding the docstring to the plot_parameters, so they can look it up in the docs

File: deepinv/unfolded/deep_equilibrium.py (hunk 2 above)
Comment (tachella) [reply deepinv/unfolded/deep_equilibrium.py]: The input type, i.e., bool

:param str, torch.nn.Module iterator: ....
:param bool g_first: ...

File: examples/plug-and-play/demo_PnP_custom_optim.py (hunk 9 above)
Comment (tachella) [reply examples/plug-and-play/demo_PnP_custom_optim.py:5]: done

File: examples/unfolded/demo_custom_prior_unfolded.py [hunk 17]
```diff
@@ -129,20 +128,21 @@ def nabla(I):
     return G
 
 
-# Define the prior
+# Define the smooth TV prior with a Huber loss.
```
Comment (matthieutrs) [inline examples/unfolded/demo_custom_prior_unfolded.py:131]: Not true anymore

**Review by matthieutrs — COMMENTED** [review]
Great job thanks! Remaining 1 minor update of doc on Huber loss deprecated that I'll do now.
